clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -406,6 +406,55 @@ def DataMemberAttr : CIR_Attr<"DataMember", "data_member",
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// MethodAttr
+//===----------------------------------------------------------------------===//
+
+def MethodAttr : CIR_Attr<"Method", "method", [TypedAttrInterface]> {
+  let summary = "Holds a constant pointer-to-mem-func to non-virtual function";
+  let description = [{
+    A method attribute is a literal attribute that represents a constant
+    pointer-to-member-function value which points to a non-virtual member
+    function.
+
+    The `symbol` parameter gives the global symbol for the non-virtual member
+    function.
+  }];
+
+  let parameters = (ins AttributeSelfTypeParameter<
+                            "", "mlir::cir::MethodType">:$type,
+                        OptionalParameter<
+                            "std::optional<FlatSymbolRefAttr>">:$symbol);
+
+  let assemblyFormat = [{
+    `<` ($symbol^):(`null`)? `>`
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// VirtualMethodAttr
+//===----------------------------------------------------------------------===//
+
+def VirtualMethodAttr : CIR_Attr<"VirtualMethod", "virtual_method",
+                                 [TypedAttrInterface]> {
+  let summary = "Holds a constant pointer-to-mem-func to virtual function";
+  let description = [{
+    A virtual method attribute is a literal attribute that represents a constant
+    pointer-to-member-function value which points to a virtual member function.
+
+    The `vtable_offset` parameter gives the offset of the vtable entry
+    corresponding to the virtual member function represented by this attribute.
+  }];
+
+  let parameters = (ins AttributeSelfTypeParameter<
+                            "", "mlir::cir::MethodType">:$type,
```

Comment:
Can we encode if it's virtual in the `MethodType` instead? One `MethodAttr` seems enough. It should then have two optionals and verifier ensure some of the semantics.

Reply:
Updated.